sklearn/cluster/_kmeans.py

```diff
@@ -926,6 +946,32 @@ class KMeans(
            [ 1.,  2.]])
     """
 
+    _expected_params_type_and_vals = {
+        "n_clusters": [(numbers.Integral, Interval(1, None, closed="left"))],
```

Comment:
In this version, `_validate_params` does not handle valid values that depend on the data, like n_clusters should be <= n_samples.

To extend this PR, we can imagine calling `_validate_params` after `_validate_data` and accepting specific strings in the Interval bounds like "n_samples", "n_features" that are resolved at check time.